File: Исходники/Глава 2. Часть 1/Разделяй и властвуй/3. Сортировка слиянием/C++/Олимпиадный/MergeSort.cpp
#include <vector>
#include <iostream>
#include <ctime>


std::vector<std::vector<int>> split_array(std::vector<int> &array) {
    std::vector<std::vector<int>> single_array;
    for (int i : array) {
        single_array.push_back({ i });
    }
    return single_array;
}


std::vector<int> sort_arrays(std::vector<int> &array_1, std::vector<int> &array_2) {
    std::vector<int> first = array_1;
    std::vector<int> second = array_2;
    int i = 0;
    int j = 0;
    std::vector<int> new_array;
    while (i < first.size() and j < second.size()) {
        if (first[i] < second[j]) {
            new_array.push_back(first[i]);
            i++;
        }
        else {
            new_array.push_back(second[j]);
            j++;
        }
    }
    if (i < first.size() or j < second.size()) {
        new_array.insert(new_array.end(), first.begin() + i, first.end());
        new_array.insert(new_array.end(), second.begin() + j, second.end());
    }
    return new_array;
}


std::vector<int> merge(std::vector<std::vector<int>> &array) {
    std::vector<std::vector<int>> final_array;
    if (array.size() % 2 > 0) {
        array.push_back({});
    }
    for (int i = 0; i < array.size() - 1; i += 2) {
        std::vector<int> curr = sort_arrays(array[i], array[i + 1]);
        final_array.push_back(curr);
    }
    if (final_array.size() == 1) {
        return final_array[0];
    }
    return merge(final_array);
}


std::vector<int> merge_sort(std::vector<int>& array) {
    std::vector<std::vector<int>> single_array = split_array(array);
    return merge(single_array);
}

File: Исходники/Глава 2. Часть 1/Разделяй и властвуй/3. Сортировка слиянием/Python/MergeSort.py

def array_in_half(array):
    if len(array) > 1:
        middle = len(array) // 2
        return [array[:middle], array[middle:]]
    return array


def split_array(array):
    if len(array) == 1:
        return [array]
    first, second = array_in_half(array)
    return split_array(first) + split_array(second)


def sort_arrays(arrays):
    first, second = arrays
    i = j = 0
    new_array = []
    while i < len(first) and j < len(second):
        if first[i] < second[j]:
            new_array.append(first[i])
            i += 1
        else:
            new_array.append(second[j])
            j += 1
    if i < len(first) or j < len(second):
        new_array += first[i:len(first)]
        new_array += second[j:len(second)]
    return new_array


def merge(array):
    final_array = []
    if len(array) % 2 > 0:
        array.append([])
    for i in range(0, len(array)-1, 2):
        curr = sort_arrays([array[i], array[i + 1]])
        final_array.append(curr)
    if len(final_array) == 1:
        return final_array[0]
    return merge(final_array)


def merge_sort(array):
    array = split_array(array)
    return merge(array)




